ROX-11172: document oss helm chart usage

jschnath opened the pull request:
## Description

Documentation regarding https://github.com/stackrox/stackrox/pull/1864

## Checklist
- ~[ ] Investigated and inspected CI test results~
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- [ ] Documented user facing changes (updated README.md)

If any of these don't apply, please comment below.

## Testing Performed

msugakov commented on README.md:
```diff
@@ -52,6 +53,14 @@ to https://localhost:8000/. Credentials for the 'admin' user can be found in
 `./deploy/k8s/central-deploy/password`
 (`deploy/openshift/central-deploy/password` in the OpenShift case).
 
+### Helm Charts
+
+Alternatively, there are also opensource flavor [Helm Charts](https://helm.sh/) available at [github.com/stackrox/helm-charts](https://github.com/stackrox/helm-charts/tree/main/opensource) which can be used to install StackRox
```
I appreciate you starting this, however, I believe, this should be done differently in order to bring real benefit for community users.
I described the idea [here](https://github.com/stackrox/stackrox/pull/1864#issuecomment-1141003041):
> Remember, our goal is to make entry barrier as low as possible for community users. This README can follow a typical pattern of open source projects such as first introduce the product (you don't have to write that), next explain how to install it (your part), next explain how to develop for it (currently existing content).

If you read the `## Deployment` section above, it will become clear that it assumes the installing user has to clone the source code of this repo.
a) This is already quite an entry barrier to require _just any user_ to install git and clone the repo in order to install the product.
b) This deploys just the current revision and not necessarily the stable release.

I'd ask you to try impersonate a stranger user who wants to spend as less time as possible to get the product installed and working for them. This means we need to move Helm installation to be the first thing, and, actually, the only one in the context of `## Quickstart installation`.

Secondly, instructions for secured cluster services are missing. Please add that too, otherwise the product isn't deployed completely. You may learn how to do that by following the official instructions yourself, see https://docs.openshift.com/acs/installing/installing_helm/install-helm-quick.html

Next, these instructions will likely not work normally until there are official release charts published to the repo. Therefore, merging this PR now will lead to more confusion than actually help users. Let's think how we can publish opensource charts for 3.70.0 to that location (another JIRA task to be created). Otherwise, we need to wait until everything happens automatically as part of 3.71.0, in 9 weeks from now.

Finally, I'd ask you to provide a link at the end of this section to the official online product docs for users who want to learn more about customizing Helm installation: https://docs.openshift.com/acs/installing/installing_helm/install-helm-customization.html

jschnath replied:
Fair enough, I'll rework this.

msugakov reviewed (changes requested):
This round I have mostly cosmetics. Next rounds I may find more and I also want to try run these commands myself, therefore please expect more comments in the future.

Also, once we finish iterating on this PR between us, I suggest to invite Docs Team members so they can also review and improve this user-facing part of documentation.

jschnath commented on README.md:
````diff
@@ -32,9 +33,45 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+```
+To see all available Helm charts in the repo run
+```
+helm search repo stackrox --devel
````
I'll remove it then

jschnath commented on README.md:
````diff
@@ -32,9 +33,45 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+```
+helm search repo stackrox --devel
+```
+From here you can install stackrox-central-services to install the centralized components (Central and Scanner)
````
Yeah this is clearer

jschnath commented on README.md:
````diff
@@ -32,9 +33,45 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+```
+To create a secured cluster you first need to create an init bundle containing all of the secrets.
+```
+kubectl -n rhacs-operator exec deploy/central -- roxctl --insecure-skip-tls-verify \
````
I'm not sure what you mean, what are the s and g at the end?

msugakov replied:
Had a chat in Slack.

jschnath commented on README.md:
````diff
@@ -32,9 +33,45 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+    -f ~/tmp/cli-bundle.yaml \ 
+    --set clusterName=<name_of_the_secured_cluster> \
+```
+While specifying the path to the init bundle created in the previous step, and the name of the secured cluster. When creating clusters hosted elsewhere, you will also need to specify the endpoint (address and port number) for Central.
````
Ok

jschnath commented on README.md:
````diff
@@ -32,9 +33,45 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+```
+While specifying the path to the init bundle created in the previous step, and the name of the secured cluster. When creating clusters hosted elsewhere, you will also need to specify the endpoint (address and port number) for Central.
+
+To further customize your Helm installation consult this [document](https://docs.openshift.com/acs/3.69/installing/installing_helm/install-helm-customization.html).
````
Ok

jschnath commented:
Thanks for the feedback, i've made the changes you suggested. I also pushed Helm Charts for 3.70.0 to our helm-charts repo so those are available to you for testing.

msugakov commented on README.md:
```diff
@@ -32,9 +33,44 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+## Quick Installation via Helm
 
-To quickly deploy the latest development version of StackRox to your kubernetes
+StackRox offers quick installation via Helm Charts, follow the [Helm Installation Guide](https://helm.sh/docs/intro/install/) for system specific instructions and requirements.
```
```suggestion
StackRox offers quick installation via Helm Charts. Follow the [Helm Installation Guide](https://helm.sh/docs/intro/install/) to get `helm` CLI on your system.
```

jschnath replied:
Done

msugakov replied:
Well, I still see the old text.

jschnath replied:
Done

msugakov commented on README.md:
```diff
@@ -32,9 +33,44 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+To further customize your Helm installation consult these documents:
+* <https://docs.openshift.com/acs/installing/installing_helm/install-helm-quick.html>
+* <https://docs.openshift.com/acs/installing/installing_helm/install-helm-customization.html>
+## Manual Deployment
```
Please add blank line before the next section.
```suggestion

## Manual Deployment
```

I use Markdown Linter in VS Code to help me write "more canonical" documents, and having blank line before and after `####` sections is one of the rule it suggests.

jschnath replied:
I'll see if I can find a markdown linter for Goland then

msugakov commented on README.md:
````diff
@@ -32,9 +33,44 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+```sh
+helm install -n stackrox-oss --create-namespace stackrox-central-services stackrox-oss/stackrox-central-services --devel
+```
+To create a secured cluster you first need to create an init bundle containing all of the secrets.
````
```suggestion
To create a secured cluster, you first need to generate an init bundle containing initialization secrets.
```

jschnath replied:
Done

msugakov commented on README.md:
````diff
@@ -32,9 +33,44 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+kubectl -n stackrox exec deploy/central -- roxctl --insecure-skip-tls-verify \
+    --password "$(kubectl -n stackrox get secret central-htpasswd -o go-template='{{index .data "password" | base64decode}}')" \
+    central init-bundles generate stackrox-init-bundle --output - > stackrox-init-bundle.yaml
+```
````
```suggestion
```
The init bundle will be saved in `stackrox-init-bundle.yaml`. Use it to provision secured clusters as shown below.
```

jschnath replied:
Done

msugakov commented on README.md:
````diff
@@ -32,9 +33,44 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+    --password "$(kubectl -n stackrox get secret central-htpasswd -o go-template='{{index .data "password" | base64decode}}')" \
+    central init-bundles generate stackrox-init-bundle --output - > stackrox-init-bundle.yaml
+```
+Then install Secured Cluster Services in the same cluster using this command:
````
Maybe let's use "stackrox-secured-cluster-services" instead of "Secured Cluster Services" to be consistent with "stackrox-central-services" you set before.
```suggestion
Then install stackrox-secured-cluster-services in the same cluster using this command:
```

jschnath replied:
Done

msugakov commented on README.md:
```diff
@@ -32,9 +33,44 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+To further customize your Helm installation consult these documents:
+* <https://docs.openshift.com/acs/installing/installing_helm/install-helm-quick.html>
+* <https://docs.openshift.com/acs/installing/installing_helm/install-helm-customization.html>
```
At this point the installation is done, but I think we're missing something important. After the user installed StackRox, the user wants to see it. Let's add simple instructions for these first steps. Something like
```
### First use

Follow these steps to get to StackRox UI

1. Obtain admin password.
kubectl -n stackrox get secret central-htpasswd -o go-template='{{index .data "password" | base64decode}}'

2. Setup port forward to central.
kubectl -n stackrox port-forward deploy/central 8000:8443

3. Open https://localhost:8000 in your browser.

4. Accept certificate warnings and proceed.

5. Log in as `admin` using password printed by the command above.
```

jschnath replied:
Done

msugakov reviewed (commented):
Few more nitpicks and one suggestion. Still did not run commands myself, but will do.

msugakov commented on README.md:
```diff
@@ -32,9 +34,60 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+Follow these steps to get to StackRox UI
+
+1. Obtain admin password:
+   `kubectl -n stackrox get secret central-htpasswd -o go-template='{{index .data "password" | base64decode}}'`
+
+2. Setup port forward to central:
+   `kubectl -n stackrox port-forward deploy/central 8000:8443`
```
I suppose this renders not how you wanted it to look. See https://github.com/stackrox/stackrox/tree/jschnath/ROX-11172_Document_OSS_Helm_Chart_usage#first-use

I think you need to use triple backticks around code not single ones if you'd like code blocks to be rendered as blocks.

If you prefer the code to be rendered inline like how it is now, that's also fine by me.

----

Screenshot of the current rendering
![image](https://user-images.githubusercontent.com/537715/171587002-47f910c0-696c-441a-8dba-8fd7e0839c4f.png)

jschnath replied:
backticks and triple backticks render the same in Goland, I'll put in triples here

msugakov commented on README.md:
```diff
@@ -32,9 +34,60 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+2. Setup port forward to central:
+   `kubectl -n stackrox port-forward deploy/central 8000:8443`
+
+3. Open https://localhost:8000 in your browser.
```
Nitpick: linter likes bare URLs to be surrounded with `<` and `>` characters.
```suggestion
3. Open <https://localhost:8000> in your browser.
```

jschnath replied:
ok

msugakov reviewed (commented):
Will go through commands now

msugakov commented on README.md:
````diff
@@ -32,9 +34,60 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+```
+Then install stackrox-secured-cluster-services in the same cluster using this command with the init bundle you just generated:
+```sh
+helm install -n stackrox-oss --create-namespace stackrox-secured-cluster-services stackrox-oss/secured-cluster-services \
+    -f stackrox-init-bundle.yaml \ 
+    --set clusterName=<name_of_the_secured_cluster> 
````
1. Couple fixes
```suggestion
helm install -n stackrox stackrox-secured-cluster-services stackrox-oss/stackrox-secured-cluster-services \
    -f stackrox-init-bundle.yaml \
    --set clusterName=<name_of_the_secured_cluster> 
```

2. Again, for development/small clusters it make sense to update resources to be more conservative.
See development values here https://github.com/stackrox/stackrox/blob/3149711ad41787b732c6afb674e5384848af6c9b/deploy/common/k8sbased.sh#L533

However, unlike updating central services after they were deployed, upgrading secured-cluster-services is finicky. Better to append request values upon creation. Here's what I came up with:

```
        --set sensor.resources.requests.memory=500Mi \
        --set sensor.resources.requests.cpu=500m \
        --set sensor.resources.limits.memory=500Mi \
        --set sensor.resources.limits.cpu=500m
```

You can say something like:

> Append the following arguments to the above command in order to limit sensor resources usage if you're deploying on development or small clusters. Do not use these values on production clusters.

msugakov commented on README.md:
````diff
@@ -32,9 +34,60 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+Follow these steps to get to StackRox UI
+
+1. Obtain admin password:
+   ```kubectl -n stackrox get secret central-htpasswd -o go-template='{{index .data "password" | base64decode}}'```
````
For things to render correctly, triple backticks should be each on separate line.

jschnath commented:
I just reran through the entire process with a minikube and limited resources. I think I've also addressed all suggestions.

<img width="519" alt="Screenshot 2022-06-03 at 13 46 21" src="https://user-images.githubusercontent.com/102952141/171847910-43eebbe5-ca65-4fdf-be8d-3b2e4fa1b77b.png">

msugakov commented on README.md:
````diff
@@ -39,32 +39,52 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
 ```
 Make sure to provide some name in `clusterName` argument meaningful to you. The cluster will be identified by this name in clusters list in StackRox UI.
 
 When deploying Secured Cluster Services on a different cluster, you will also need to specify the endpoint (address and port number) of Central via `--set centralEndpoint=<endpoint_of_central_service>` command-line argument.
````
Please add something like:

> When deploying StackRox on a small node, add the following arguments to the `helm install` command above. This should reduce stackrox-secured-cluster-services resource requirements. Don't include these when deploying on a production cluster.
```
        --set sensor.resources.requests.memory=500Mi \
        --set sensor.resources.requests.cpu=500m \
        --set sensor.resources.limits.memory=500Mi \
        --set sensor.resources.limits.cpu=500m
```

jschnath replied:
Done

jschnath commented on README.md:
````diff
@@ -39,32 +39,52 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
 ```
 Make sure to provide some name in `clusterName` argument meaningful to you. The cluster will be identified by this name in clusters list in StackRox UI.
 
 When deploying Secured Cluster Services on a different cluster, you will also need to specify the endpoint (address and port number) of Central via `--set centralEndpoint=<endpoint_of_central_service>` command-line argument.
````
Done

jschnath commented on README.md:
````diff
@@ -39,32 +39,52 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+  --set scanner.resources.requests.cpu=500m \
+  --set scanner.resources.limits.memory=2500Mi \
+  --set scanner.resources.limits.cpu=2000m
+```
````
Okay I moved it

jschnath commented:
I think it's all good now.

Maddosaurus commented on README.md:
```diff
@@ -32,9 +34,87 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+2.Open <https://localhost:8000> in your browser.
+
+3.Accept certificate warnings and proceed.
```
How about we leave out this step?
Depending on the setup, the user might not even encounter a cert warning (e.g. when adding their own CA in the setup process).
Also, we're kind of motivating users to do something potentially dangerous here (click away this warning).
I would hope that users are aware of the implications, so what do you think?

jschnath replied:
Okay I can leave it out but in that case I think we should also do what we can to avoid creating those warnings in the first place.

kcarmichael08 replied:
Suggestion:
If a certificate warning is displayed, you can accept the warnings to install the default self-signed certificate. To configure custom certificates, see the section on "Adding Custom Certificates" in [docs.openshift.com/acs/configuration/add-custom-certificates.html](https://docs.openshift.com/acs/configuration/add-custom-certificates.html).

jschnath replied:
Went with Michaels suggestion

Maddosaurus reviewed (approved):
I have some small nitpicks regarding wording, but overall, solid documentation of the Helm chart setup. Well done! 😸

jschnath commented on README.md:
```diff
@@ -32,9 +34,87 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+When deploying stackrox-secured-cluster-services on a different cluster, you will also need to specify the endpoint (address and port number) of Central via `--set centralEndpoint=<endpoint_of_central_service>` command-line argument.
+
+When deploying StackRox on a small node, add the following arguments to the helm install command above. This should reduce stackrox-secured-cluster-services resource requirements. Don't include these when deploying on a production cluster.
```
I'll rephrase this one the same as above then

msugakov commented:
Hi @kcarmichael08 @gaurav-nelson and @lmaynard88 

Could one of you please review documentation instructions in this PR and provide your feedback?
The idea is that this is top-level README file in the repo and we want to give community users the most straightforward way to install a community build of StackRox.

msugakov commented on README.md:
```diff
@@ -108,9 +108,7 @@ kubectl -n stackrox port-forward deploy/central 8000:8443
 
 2.Open <https://localhost:8000> in your browser.
 
-3.Accept certificate warnings and proceed.
```
I disagree about removing this, I think we should say something about self-signed certs.

Maddosaurus replied:
How about we rephrase it? Something along the lines of:
"In a quick setup, you will encounter a self-signed certificate here, which you must accept to continue. For a custom certificate setup, consult https://docs.openshift.com/acs/configuration/add-custom-certificates.html"?
What do you think?

jschnath replied:
Went with the suggestion of Michael in the end, I think this should be good

msugakov commented on README.md:
````diff
@@ -32,9 +34,87 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+helm install -n stackrox --create-namespace stackrox-central-services stackrox/stackrox-central-services --set central.adminPassword.value="$(cat stackrox-admin-password.txt)"
+```
+
+If you're deploying StackRox on a small node like a local development cluster, run the following command to reduce StackRox resource requirements. Please keep in mind that these reduced resource settings are not suited for production setups.
````
```suggestion
If you're deploying StackRox on a small node like a local development cluster, run the following command to reduce StackRox resource requirements. Please keep in mind that these reduced resource settings are not suited for production setup.
```

jschnath replied:
That would make it grammatically incorrect. I can change it to "not suited for a production setup" instead.

msugakov replied:
Sure!

msugakov commented on README.md:
````diff
@@ -32,9 +34,87 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+Then install stackrox-secured-cluster-services in the same cluster using this command with the init bundle you just generated:
+```sh
+helm install -n stackrox stackrox-secured-cluster-services stackrox/stackrox-secured-cluster-services \
+  -f stackrox-init-bundle.yaml \ 
````
Please do not leave whitespace after `\` character, this breaks the command.

jschnath replied:
Good to know, I was wondering why it broke last time I copied it over but didn't figure out why.

jschnath commented on README.md:
````diff
@@ -32,9 +34,87 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+  --set scanner.resources.limits.memory=2500Mi \
+  --set scanner.resources.limits.cpu=2000m
+```
+To create a secured cluster, you first need to generate an init bundle containing initialization secrets. The init bundle will be saved in `stackrox-init-bundle.yaml`. Use it to provision secured clusters as shown below.
````
ok

jschnath commented on README.md:
```diff
@@ -32,9 +34,87 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+When deploying stackrox-secured-cluster-services on a different cluster, you will also need to specify the endpoint (address and port number) of Central via `--set centralEndpoint=<endpoint_of_central_service>` command-line argument.
+
+When deploying StackRox on a small node, add the following arguments to the helm install command above. This should reduce stackrox-secured-cluster-services resource requirements. Please keep in mind that these reduced resource settings are not recommended for production setups.
```
same as above, I'll do "not recommended for a production setup" then

msugakov commented on README.md:
````diff
@@ -69,28 +69,35 @@ helm upgrade -n stackrox stackrox-central-services stackrox/stackrox-central-ser
   --password "$(cat stackrox-admin-password.txt)" \
   central init-bundles generate stackrox-init-bundle --output - > stackrox-init-bundle.yaml
 ```
+Set a `CLUSTER_NAME` for your secured cluster.
````
```suggestion
Set a meaningful cluster name for your secured cluster in `CLUSTER_NAME` environment variable. The cluster will be identified by this name in clusters list in StackRox UI.
```

msugakov commented on README.md:
````diff
@@ -69,28 +69,35 @@ helm upgrade -n stackrox stackrox-central-services stackrox/stackrox-central-ser
+  -f stackrox-init-bundle.yaml \
+  --set clusterName="$CLUSTER_NAME"
 ```
 Make sure to provide some name in `clusterName` argument meaningful to you. The cluster will be identified by this name in clusters list in StackRox UI.
````
```suggestion
```

msugakov commented on README.md:
````diff
@@ -69,28 +69,35 @@ helm upgrade -n stackrox stackrox-central-services stackrox/stackrox-central-ser
 ```
 Make sure to provide some name in `clusterName` argument meaningful to you. The cluster will be identified by this name in clusters list in StackRox UI.
 
 When deploying stackrox-secured-cluster-services on a different cluster, you will also need to specify the endpoint (address and port number) of Central via `--set centralEndpoint=<endpoint_of_central_service>` command-line argument.
````
```suggestion
When deploying stackrox-secured-cluster-services on a different cluster than the one where stackrox-central-services are deployed, you will also need to specify the endpoint (address and port number) of Central via `--set centralEndpoint=<endpoint_of_central_service>` command-line argument.
```

msugakov reviewed (approved):
LGTM.

1. I [re-pinged](https://srox.slack.com/archives/CELUQKESC/p1655113132994849) Gaurav, Kerry and Lynne in Slack. Let's give them another day to respond, if they're interested to review.
2. Please check if you want to use suggestion that Matthias shared in his last comment.
3. Before finally merging this PR, please re-run commands one more time to make sure edits are still valid.

kcarmichael08 commented on README.md:
````diff
@@ -32,9 +34,92 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+helm install -n stackrox --create-namespace stackrox-central-services stackrox/stackrox-central-services --set central.adminPassword.value="$(cat stackrox-admin-password.txt)"
+```
+
+If you're deploying StackRox on a small node like a local development cluster, run the following command to reduce StackRox resource requirements. Please keep in mind that these reduced resource settings are not suited for a production setup.
````
generally avoid "please" in technical docs

jschnath replied:
Fixed those as well as other instances in the document.

kcarmichael08 commented on README.md:
```diff
@@ -32,9 +34,92 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+3.Accept certificate warnings and proceed.
+
+4.Log in as `admin` using the password in `stackrox-admin-password.txt`
```
do we need to tell users where this file is? (txt file)

jschnath replied:
Yes, we specifically generate this password file to make it more convenient for the user. The previous approach was to not specify a password and extract the automatically generated password using `kubectl -n stackrox get secret central-htpasswd -o go-template='{{index .data "password" | base64decode}}'`

But it turns out only Operator-based install write the `.data.password` attribute, so we decided to manually set a password instead. It's written into `stackrox-admin-password.txt` to be easily accessible, users are free to change that of course.

jschnath commented on README.md:
```diff
@@ -32,9 +34,92 @@ For alternative ways, stop by our Community Hub [stackrox.io](https://www.stackr
+
+## Manual Deployment
+
+To manually deploy the latest development version of StackRox to your kubernetes
```
Fixed those as well as other instances in the document.
